### DISCUSSION TITLE
Support -r flags in the pip-compile manager

### DISCUSSION THREAD

**Discussion by mbudnek** [renovatebot/renovate#28050] [Suggest an Idea]
### Tell us more.

With pip-compile it's possible to reference one requirements file from another using a `-r` flag.

For example, you could have a "requirements-base.in" that looks like this:
```
some-base-requirement==1.2.3
```

And "requirements-extra.in" that looks like this:
```
-r requirements-base.txt

some-extra-requirement==1.2.3
```
When renovate updates `some-base-requirement`, the pip-compile manager needs to update both "requirements-base.txt" and "requirements-extra.txt" even though nothing change directly in "requirements-extra.in".

---

I can see three scenarios that need to be supported:

1. An input file references another input file
2. An input file references a lock file
3. An input file references a requirements file not managed by pip-compile

Scenarios (1) and (2) should be easy enough to implement by simply adding the appropriate lock files to the `PackagFile.lockFiles` array for the file that got included.  For the above example that would mean adding "requirements-extra.txt" to the `lockFiles` array for "requirements-base.in".  That way, when `some-base-requirement` gets updated in "requirements-base.in" the manager will know to update both "requirements-base.txt" and "requirements-extra.txt".

Scenario (3) is a bit trickier.  My first thought was to simply copy all of the dependencies from the un-managed requirements file into the `PackageFile.deps` array for the managed file, but that didn't work due to a couple of related issues:
1. Renovate re-parses the requirements file after doing the version string replacement, and if it detects no changes it doesn't call `updateArtifacts` at all.
2. The pip-compile manager needs more context than it's given during the re-parse to determine what's a pip-compile managed package file and what isn't.

**Comment by mbudnek** [comment]
I created #28052 to address scenarios (1) and (2).  I'm not sure if scenario (3) is worth the effort of addressing right now.

**Comment by not7cd** [comment]
I think I was thinking about this exact scenario. I believe that some cases are handled by Renovate simply detecting versions and grouping them in a PR. This is a happy accident. So I think case 2 is handled. A reproduction repository may be needed to prove that. I may be wrong, [as I'm running a fork with this change](https://github.com/renovatebot/renovate/pull/27561).

Also, this is a part of the extract function which detects if a detected input file is also matched as output/lock file.
https://github.com/renovatebot/renovate/blob/e998c65ae34d85ce6539212197e49833849128c9/lib/modules/manager/pip-compile/extract.ts#L123-L130

**Reply by mbudnek** [reply]
I added support for cases (1) and (2) in #28052 (neither was fully covered by the existing manager).  The manager needs to know to update all of the dependent lock files when a dependency is updated.  So if a requirement in "requirements-base.in" is updated, the manager has to know to re-compile both "requirements-base.txt" and "requirements-extra.txt".

Scenario (3) is basically impossible currently as far as I can tell, since the `pip-compile` manager has no idea when a dependency managed by the `pip_requirements` manager gets updated, and so can't re-compile the dependent lock files.  The next lockfile maintenance run will see the change though; so it's still sort of covered without any changes.

**Reply by not7cd** [reply]
I think I have case (1) and (2) covered by https://github.com/renovatebot/renovate/pull/27561. But only for security alerts. And this is a happy accident, so I was a bit confused about your reasoning. 

The way it works is by creating a separate updates for "indirect" dependencies. As you can see, they are added to the package file but are disabled by default. Security update forces them to update. And because they are indirect in relation to their input file, they are handled as indirect. In `updateArtifacts` they are updated explicitly with `--upgrade-package`. This still works out because all updates happen in correct order. If we could somehow force `updateArtifacts` after updating dependency, like in your PR, it should also work out.

**Comment by not7cd** [comment]
I think I wrapped my head around it. I wouldn't call this idea support for `-r` flags, because they are parsed and to some extent respected. By that, I mean all package files and lock files are order by dependencies between them through topological sort.

This ordering of lock files is happening here:
https://github.com/renovatebot/renovate/blob/e998c65ae34d85ce6539212197e49833849128c9/lib/modules/manager/pip-compile/utils.ts#L20-L27

So if `some-base-requirement` existed in every package file and lock file, all `pip-compile` executions should happen in this order. But I understand that some of the artifact updates won't happen, as in the case with `requirements-extra.in`. I'm not sure about the solution to that, it may be as easy as adding every dependent lock file to a given package file _lock files_ in correct order. 

I have two suggestions outright:
* use `depsBetweenFiles` because it already has information from `pip_requirements` manager data alongside requirements and constraints from `pip-compile` arguments
* this feature could be renamed to something like: "artifact update propagation to dependent lock files" or something in that spirit.

**Reply by mbudnek** [reply]
My solution was to simply walk the sorted list of package files in reverse order and add extra `lockFiles` entries for the dependent files.  That works for all the cases I mentioned since `updateArtifacts` goes entirely by the lock file header to determine which input files to compile.

So in the example from the first post in this discussion, "requirements-base.in" will have a `lockFiles` array of `["requirements-base.txt", "requirements-extra.txt"]` and the manager will do the right thing and re-compile both "requirements-base.in" and "requiremetns-extra.in".

This also works for transitive dependencies.  i.e. if "requirements-more-extra.in" had `-r requirements-extra.txt` then the final `lockFiles` would be something like
```
[
  {
    packageFile: "requirements-base.in",
    lockFiles: ["requirements-base.txt", "requirements-extra.txt", "requirements-more-extra.txt"],
  },
  {
    packageFile: "requirements-extra.in",
    lockFiles: ["requirements-extra.txt", "requirements-more-extra.txt"],
  },
  {
    packageFile: "requirements-more-extra.in",
    lockFiles: ["requirements-more-extra.txt"],
  }
]
```

This does have the slight downside that if the same dependency is listed in both "requirements-base.in" and "requirements-extra.in" then the latter will get re-compiled twice.

**Reply by not7cd** [reply]
It may be a necessary downside. More reasons to not duplicate requirements across files and layer them properly.

Looking at #28052 I see space for improvement. Lock file sorting can be handled by an existing function in `utils.ts`. Because we already have a `Graph` object, maybe we can extract all children of a given node. And just check if these children are a valid lock files. All the data should already be here.

**Reply by not7cd** [reply]
https://github.com/datavis-tech/graph-data-structure, `graph.depthFirstSearch()` may do the trick

**Reply by mbudnek** [reply]
I didn't see any reason to dig through the graph when there's already a topologically-sorted array.  Since it's already topologically-sorted, simply iterating in reverse order will make everything "bubble up" in the correct, sorted order.

**Comment by not7cd** [comment]
Regarding (3). For clarification. This is a case where a file falls outside scope of detected package files?

```
-r outside.in
dep>=1.0.0
```

```
#    pip-compile --output-file=managed.txt managed.in
dep==1.0.0
```

```mermaid
graph LR
     managed.txt[[managed.txt]]
     outside.in --> managed.in
     managed.in --> managed.txt
     
```

**Reply by mbudnek** [reply]
Yes, for example you have something like a requirements-common.txt that isn't generated by pip-compile and one or more requirements-service.in files with `-r requirements-common.txt`.  In theory there's no reason to pip-compile your common requirements file since it's never installed directly, only referenced by `-r`.
